joomla/joomla-cms
# [4.0] Implement loading=lazy for images

Pull Request for Issue #27761

### Summary of Changes

- Implement a new loading=lazy attribute for images on content via a plugin that can be disabled.
- Implement a default enabled loading=lazy option that can be overwritten by extension devs useing the options array.

As discussed here: https://volunteers.joomla.org/departments/production/reports/1226-production-dept-meeting-april-21-2020

> Allow progressive features
> Joomla's current policy is that we only accept changes or features that work for everything that meets our minimum requirement (e.g. PHP versions, browsers, database versions etc..)
>
>The motion goes beyond this by allowing progressive features. This allows us to take advantage of new technology for those that use it as long as there is no adverse effect to those that don't (this should also take into account the support/documentation affects on such conditional features). As an example we could look at "lazy loading" or at "same site cookies". This motion means that a PR or ISSUE should not be closed or rejected on the argument that it IS a progressive feature. The acceptance of such an issue or feature is still subject to the usual and accepted scrutiny and discussion.

We are now even official allowed to introduce featues like this here that don't break existing sites and just improve the overall expiricece for people using modern browsers.

#### Feature

This PR sets the lazyloading attribute to all images that passes the onContentPrepare Event or HTMLHelper::image to allow modern browsers to lazyload the images.

More information about the `loading` attribute:
- Description: [https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-loading](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-loading).
- HTML Specification: [https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-loading](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-loading).
- Currently the `loading` attribute is supported in the following browsers: [https://caniuse.com/#feat=loading-lazy-attr](https://caniuse.com/#feat=loading-lazy-attr).

### Testing Instructions

- install 4.0
- setup one article with images and atd it to the frontend
- apply this PR
- discover install the plugin
- enable the plugin
- check the frontend
- see that it not has the lazyloading attribute
- disable the plugin
- see that the loading attribute is not set

### Expected result

Joomla by its core can set the loading=lazy attribute to images.

### Actual result

Joomla can not set that loading=lazy attribute.

### Documentation Changes Required

A doc page will be produced on this, I have just added the label and assigned it to me.

### Beyond this PR 

There is a backport of this feature / plugin for Joomla 3 here: [GitHub](https://github.com/zero-24/plg_content_imagelazyloading) [JED](https://extensions.joomla.org/extension/imagelazyloading/)

### Special thanks

Thanks @SniperSister and @felixarntz for your support on that PR.
